tests/test_application.py

```diff
@@ -99,6 +99,76 @@ def test_client_certificate(self):
+        with open(key_path) as f:
+            pem = f.read()
+        with open(public_certificate) as f:
+            public_certificate = f.read()
+        app = ConfidentialClientApplication(
+            CONFIG['client_id'],
+            client_credential={"private_key": pem, "thumbprint": CONFIG["thumbprint"],
+                               "public_certificate": public_certificate})
+        x5c = app._extract_x5c_value()
+        self.assertIsInstance(x5c, list)
+        self.assertEquals(len(x5c), 1)
+
+    @unittest.skipUnless("public_certificate" in CONFIG, "Missing Public cert")
+    def test_subject_name_issuer_authentication_input_with_begin_certificate_and_end_certificate_multiple(self):
+        assert ("private_key_file" in CONFIG
+                and "thumbprint" in CONFIG and "public_certificate" in CONFIG)
+        key_path = os.path.join(THIS_FOLDER, CONFIG['private_key_file'])
+        public_certificate = os.path.join(THIS_FOLDER, CONFIG['public_certificate'])
+        with open(key_path) as f:
+            pem = f.read()
+        with open(public_certificate) as f:
+            public_certificate = f.read()
+        app = ConfidentialClientApplication(
+            CONFIG['client_id'],
+            {"private_key": pem, "thumbprint": CONFIG["thumbprint"], "public_certificate": public_certificate})
+        x5c = app._extract_x5c_value()
+        self.assertIsInstance(x5c, list)
+        self.assertGreater(len(x5c), 1)
+
+    @unittest.skipUnless("public_certificate" in CONFIG, "Missing Public cert")
+    def test_subject_name_issuer_authentication_input_without_begin_certificate_and_end_certificate(self):
+        assert ("private_key_file" in CONFIG
+                and "thumbprint" in CONFIG and "public_certificate" in CONFIG)
+        key_path = os.path.join(THIS_FOLDER, CONFIG['private_key_file'])
+        public_certificate = os.path.join(THIS_FOLDER, CONFIG['public_certificate'])
+        with open(key_path) as f:
+            pem = f.read()
+        with open(public_certificate) as f:
+            public_certificate = f.read()
+        app = ConfidentialClientApplication(
+            CONFIG['client_id'],
+            {"private_key": pem, "thumbprint": CONFIG["thumbprint"], "public_certificate": public_certificate})
+        x5c = app._extract_x5c_value()
+        self.assertIsInstance(x5c, list)
+        self.assertEquals(len(x5c), 1)
+
+    @unittest.skipUnless("public_certificate" in CONFIG, "Missing Public cert")
+    def test_subject_name_issuer_authentication(self):
+        assert ("private_key_file" in CONFIG
+                and "thumbprint" in CONFIG and "public_certificate" in CONFIG)
+        key_path = os.path.join(THIS_FOLDER, CONFIG['private_key_file'])
+        public_certificate = os.path.join(THIS_FOLDER, CONFIG['public_certificate'])
+        with open(key_path) as f:
+            pem = f.read()
+        with open(public_certificate) as f:
```

Comment:
Rather than defining a variable just 3 lines above and then override its content (thus its meaning) 1 line below, let's just do this, thus avoiding one of the [most difficult things in computer science](https://image.slidesharecdn.com/naming-3-170610132552/95/2nd-hardest-problem-in-computer-science-3-638.jpg?cb=1497101264) :-)

```python
with open(os.path.join(...)) as f:
    public_certificate = f.read()
```

PS: The `key_path` and `pem` variables did not have such naming conflict, but you can go ahead to also apply the same technique. But we do NOT need to retrofit this new technique to pre-existing test cases.